### PR TITLE
IGNITE-22633 Catalog compaction. Implement choosing of coordinator

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-22633

------------

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests to verify choosing of coordinator for catalog compaction.
+ */
+class ItCatalogCompactionCoordinatorChoosingTest extends ClusterPerClassIntegrationTest {
+    private static final int CLUSTER_SIZE = 5;
+    private static final int MINORITY_SIZE = (CLUSTER_SIZE - 1) / 2;
+
+    // test is sensible to count of nodes in custer
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java]: fixed

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageLeaderElectionListener.java [hunk 2]
```diff
@@ -91,18 +93,22 @@ public class MetaStorageLeaderElectionListener implements LeaderElectionListener
...
         this.logicalTopologyService = logicalTopologyService;
         this.metaStorageSvcFut = metaStorageSvcFut;
         this.clusterTime = clusterTime;
         this.metaStorageConfigurationFuture = metaStorageConfigurationFuture;
+        this.electionListeners = electionListeners;
     }
 
     @Override
     public void onLeaderElected(ClusterNode node, long term) {
+        electionListeners.forEach(listener -> listener.onLeaderElected(node));
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageLeaderElectionListener.java:110]: good point, added line to ElectionListener class-level javadoc

File: modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            String coordinatorNodeId = ((CatalogManagerImpl) node.catalogManager()).compactionCoordinator();
+
+            if (coordinatorNodeId == null) {
+                coordinatorHasNotBeanUpdatedCount++;
+            } else {
+                coordinators.add(coordinatorNodeId);
+            }
+        }
+
+        assertThat(coordinatorHasNotBeanUpdatedCount, lessOrEqualTo(MINORITY_SIZE));
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java:69]: not exactly. In this test metastorage group consists of all nodes of cluster, but leader can be elected as soon as majority get to consensus. Given that said, it's possible, that leader will be elected when majority of the node started. If minority of the nodes join the cluster _after_ leader had been elected, they won't receive notification, thus `compactionCoordinatorNodeName` on such nodes may remain `null`. This assertion makes sure, that expectations are true, and election happens as soon as (at least) majority of the node started.

**Review by xtern — APPROVED** [review]

File: modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    void coordinatorIsChosenAfterStartAndReelectedAfterNodeCrash() throws Exception {
+        Set<String> coordinators = new HashSet<>();
+
+        int coordinatorHasNotBeanUpdatedCount = 0;
+        for (IgniteImpl node : runningNodes()) {
+            String coordinatorNodeId = ((CatalogManagerImpl) node.catalogManager()).compactionCoordinator();
+
+            if (coordinatorNodeId == null) {
+                coordinatorHasNotBeanUpdatedCount++;
```
Comment (lowka) [inline modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java:63]: There is a typo:
> coordinatorHasNotBe_a_nUpdatedCount
> coordinatorHasNotBeenUpdatedCount

**Review by lowka — APPROVED** [review]

File: modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java (hunk 4 above)
Comment (korlov42) [reply modules/catalog/src/integrationTest/java/org/apache/ignite/internal/catalog/ItCatalogCompactionCoordinatorChoosingTest.java]: fixed, thanks

**Pull request merged**